### ISSUE TITLE
Unstable panel widths with more than three panels

### ISSUE THREAD

**Issue by functoria** [bvaughn/react-resizable-panels#323]
Hi!

When a panel group has more than three panels, when resizing one of the "center" panels, I've noticed the sizes of the preceding panels are not changing smoothly, but seem to "jump around". I suppose this is caused by a rounding error...

Can this be remedied as the visual experience is not great, especially when the panel content is rich and depends on the panel size? Here's a video showing the issue with a horizontal layout:

https://github.com/bvaughn/react-resizable-panels/assets/24974959/7cb20065-3c94-4f11-988a-d534f7990e3d

Thank you for the great libraries!

**Comment by bvaughn** [comment]
Can you share a repro of this please? I suspect the reason for the “jumping” is that something in your app is slow to render on resize.

**Comment by bvaughn** [comment]
Looks fine for me, even with 6x CPU throttling enabled. 🤷🏼

**Comment by functoria** [comment]
This is strange, I'm using a MacBook Pro M2, without any throttling, and I can still reproduce the effect in the video below on all my browsers:

https://github.com/bvaughn/react-resizable-panels/assets/24974959/f2573964-26cb-433b-8f02-d0c4c14808a1

(I've updated the [sandbox](https://codesandbox.io/p/sandbox/react-resizable-panels-forked-kcrnc3?file=%2Fsrc%2FApp.js%3A81%2C44).)

**Comment by bvaughn** [comment]
I dunno. Seems fine for me?
https://github.com/bvaughn/react-resizable-panels/assets/29597/2155fef6-5a8c-4218-b388-549b0e37267a

**Comment by bvaughn** [comment]
Resizing seems smooth for me?

https://github.com/bvaughn/react-resizable-panels/assets/29597/2155fef6-5a8c-4218-b388-549b0e37267a

I think I initially misunderstood what you were reporting though. I assumed the "jumping" was because of slow renders and React not keeping up with the resize updates, but I think you're talking about some small changes in percentage sizes that cause the panels to shift a bit unexpectedly during resize.

I haven't noticed that either, personally.

**Comment by functoria** [comment]
It's definitely noticeable with the right panel content (just tested on other OS/machine as well). 

And yes, I think is related to some non-determinism in how the sizes are computed [here](https://github.com/bvaughn/react-resizable-panels/blob/main/packages/react-resizable-panels/src/utils/adjustLayoutByDelta.ts), probably due to the "fuzzy" number comparisons. When slowly resizing a panel, on my machine, I've noticed some of the panels, which shouldn't change size at all yet, still have their size "jumping" around the "right" value by couple of pixels.

I'm curious if anybody else is able to reproduce the issue; or maybe the effect is so negligible for most of the use cases that nobody really notices it...

Anyway, thank you for the quick response!

**Comment by bvaughn** [comment]
Maybe you could [record a Replay](https://replay.io/) of this happening?

Does seem like it could be related to the fuzzy number / precision stuff. :(

**Comment by julianZeitler** [comment]
> It's definitely noticeable with the right panel content (just tested on other OS/machine as well).
> 
> And yes, I think is related to some non-determinism in how the sizes are computed [here](https://github.com/bvaughn/react-resizable-panels/blob/main/packages/react-resizable-panels/src/utils/adjustLayoutByDelta.ts), probably due to the "fuzzy" number comparisons. When slowly resizing a panel, on my machine, I've noticed some of the panels, which shouldn't change size at all yet, still have their size "jumping" around the "right" value by couple of pixels.
> 
> I'm curious if anybody else is able to reproduce the issue; or maybe the effect is so negligible for most of the use cases that nobody really notices it...
> 
> Anyway, thank you for the quick response!

Seems like this is the same problem as in the issue I opened a few days ago. I also put up a [sandbox](https://codesandbox.io/p/sandbox/jolly-turing-m482nz?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clty8j5ov00062v6be0k77sq6%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clty8j5ov00022v6bmy6zeg26%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clty8j5ov00032v6big37opiq%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clty8j5ov00052v6bpev3ynbe%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B42.88848018085331%252C57.11151981914669%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clty8j5ov00022v6bmy6zeg26%2522%253A%257B%2522id%2522%253A%2522clty8j5ov00022v6bmy6zeg26%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clty8j5ov00052v6bpev3ynbe%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clty8j5ov00042v6b1zb5vgo5%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clty8j5ov00052v6bpev3ynbe%2522%252C%2522activeTabId%2522%253A%2522clty8j5ov00042v6b1zb5vgo5%2522%257D%252C%2522clty8j5ov00032v6big37opiq%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clty8j5ov00032v6big37opiq%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) where I was able to duplicate that behavior (You have to play around a bit to notice it, but in some positions of the right bar, it flickers if you resize with the left bar).

**Comment by functoria** [comment]
Hi, sorry for the late come-back. 

Here's the updated [sandbox](https://codesandbox.io/p/sandbox/react-resizable-panels-forked-kcrnc3?file=%2Fsrc%2FApp.js%3A99%2C40) and a recording of the unstable panel width behavior:

https://github.com/bvaughn/react-resizable-panels/assets/24974959/335dc3a0-1804-40a2-ae21-3900e9c28ea3

As you may see when only resizing between Panel#1 and Panel#2  the PIXEL width of Panel#3 and Panel#4 are changing although they should not (none of the min sizes are reached yet). The percent sizes are indeed stable.

In my usecase I'm "observing" the panel PIXEL widths (with a resize observer) and rendering other (expensive) elements based on those widths. It would definitely help not having **pixel widths of unaffected panels changing unnecesarily** as in the example above.

I could help by looking at the code, if there is not an obvious immediate solution to fix this.

Thank you!

**Comment by bvaughn** [comment]
I *think* this is just a quirk in how flex box layout works. If you observe the CSS styles for each panel, you'll see that the `flex-grow` style values don't change for the "flickering" panels.

https://github.com/bvaughn/react-resizable-panels/assets/29597/955854fc-8353-407b-9891-0f69944149af

I don't think I can do anything to fix this, short of maybe a major overhaul in the way this library works, and I don't have the capacity for that. (This is just a pro bono side project.)

If you dig into this and have a proposal, please submit a PR. Otherwise I don't think this is something I can tackle.

**Comment by bvaughn** [comment]
If you update your test app to the following, it makes this a bit easier to see I think:
```ts
export default function App() {
  return (
    <div className={styles.PanelGroupWrapper}>
      <PanelGroup className={styles.PanelGroup} direction="horizontal">
        <PanelWithDebugInfo />
        <ResizeHandle className={styles.ResizeHandle} />
        <PanelWithDebugInfo />
        <ResizeHandle className={styles.ResizeHandle} />
        <PanelWithDebugInfo />
        <ResizeHandle className={styles.ResizeHandle} />
        <PanelWithDebugInfo />
      </PanelGroup>
    </div>
  );
}

function PanelWithDebugInfo() {
  const ref = useRef<HTMLDivElement>(null);

  useLayoutEffect(() => {
    const element = ref.current;
    if (element) {
      const update = () => {
        const { width } = element.getBoundingClientRect();
        const size = element.parentElement!.style.flexGrow;

        const percentage = element.querySelector(
          '[data-test-name="percentage"]'
        );
        if (percentage) {
          percentage.textContent = `${size}%`;
        }

        const pixels = element.querySelector('[data-test-name="pixels"]');
        if (pixels) {
          pixels.textContent = `${width}px`;
        }
      };

      update();

      const observer = new ResizeObserver(update);
      observer.observe(element);

      return () => {
        observer.disconnect();
      };
    }
  }, []);

  return (
    <Panel className={styles.PanelColumn}>
      <div className={styles.Centered} ref={ref}>
        <center>
          <div data-test-name="percentage">0%</div>
          <div data-test-name="pixels">0px</div>
        </center>
      </div>
    </Panel>
  );
}
```

https://github.com/bvaughn/react-resizable-panels/assets/29597/52c49b8e-3f72-43e8-915d-43d92e6d3286

**Comment by bvaughn** [comment]
I think I may have figured out a way to fix it. Stay tuned.

**Comment by bvaughn** [comment]
Okay check out react-resizable-panels@2.0.16. I think it should fix this for you!

---

❤️ → ☕ [givebrian.coffee](http://givebrian.coffee/)

**Comment by functoria** [comment]
Thank you for fixing this!

**Comment by Georgegriff** [comment]
@bvaughn it seems like this only was fixed in 2.0.16 and regressed immediately after, i checked this in the[ code sandbox ](https://codesandbox.io/p/sandbox/react-resizable-panels-forked-kcrnc3?file=%2Fsrc%2FApp.js%3A99%2C40) by changing the version to 2.0.16 and its fine, then 2.0.17 (and latest) its not

The change in 2.0.16 ->  https://github.com/bvaughn/react-resizable-panels/compare/2.0.15...2.0.16 packages/react-resizable-panels/src/utils/adjustLayoutByDelta.ts

<img width="1685" height="437" alt="Image" src="https://github.com/user-attachments/assets/32df4bbd-9c42-44c2-9bf4-3ab05f6fc9ae" />

Current code doesn't use toFixed
https://github.com/bvaughn/react-resizable-panels/blob/main/packages/react-resizable-panels/src/utils/adjustLayoutByDelta.ts#L203

This commit seems to be missing from the repo https://github.com/bvaughn/react-resizable-panels/commit/379cb277aaba089cb6bbc86c4171e52d93706efa
